### PR TITLE
Deprecate invoking commands in the old (non-AbstractCommand) style

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
We've ported all the Homebrew-maintained `brew` commands to use instances descended from the `AbstractCommand` class. The ultimate goal is to no longer need to inspect the call stack to determine which command is being invoked, and to reduce the number of paths through the CLI parser. However, there are still third-party commands that need to be ported (a starting point could be [this](https://github.com/search?q=Homebrew%3A%3ACLI%3A%3AParser.new+do+NOT+is%3Afork&type=code) search).

This PR adds a notice to alert when deprecated-style commands are invoked. Here is an example, where I reverted "brew info" to the prior style:

```
$ brew info
Warning: Calling `brew info'. This command needs to be refactored, as it is written in a style that is deprecated! Use inheritance from `Homebrew::AbstractCommand' ( see https://docs.brew.sh/External-Commands ) instead.
103 kegs, 53,864 files, 1.8GB
```

It's the best I could come up with given the constraints of `odeprecated`, happy to update the messaging if other folks have suggestions. (Also, this is my first time adding a deprecation notice, not sure if a (or which) date should be included.)

### CODE REVIEW

**Review by carlocab — APPROVED** [review]
Thanks! Can we check that this also prints the message about reporting to the correct external tap?

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me! 👍🏻ing @carlocab's request before this is (self) merged.

**Comment by Bo98** [comment]
Usually we comment out the deprecation and then enable it on the next minor release.

File: Library/Homebrew/cli/parser.rb [hunk 1]
```diff
@@ -167,6 +167,11 @@ def initialize(cmd = nil, &block)
           @command_name = T.let(T.must(cmd_location.label).chomp("_args").tr("_", "-"), String)
           @is_dev_cmd = T.let(T.must(cmd_location.absolute_path).start_with?(Commands::HOMEBREW_DEV_CMD_PATH),
                               T::Boolean)
+          odeprecated(
+            "`brew #{@command_name}'. This command needs to be refactored, as it is written in a style that",
+            "inheritance from `Homebrew::AbstractCommand' ( see https://docs.brew.sh/External-Commands )",
+            disable_for_developers: false,
+          )
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/parser.rb:174]:
```suggestion
          # odeprecated(
          #   "`brew #{@command_name}'. This command needs to be refactored, as it is written in a style that",
          #   "inheritance from `Homebrew::AbstractCommand' ( see https://docs.brew.sh/External-Commands )",
          #   disable_for_developers: false,
          # )
```

as @Bo98 mentioned.

Not sure how I missed this first time, good catch @Bo98 😓

**Comment by reitermarkus** [comment]
If we want the public API to be inheritance-based, everything inside and including `AbstractCommand` itself needs to be documented (if it isn't already) and marked `@api public`.

**Comment by dduugg** [comment]
> If we want the public API to be inheritance-based, everything inside and including `AbstractCommand` itself needs to be documented (if it isn't already) and marked `@api public`.

Just checking that we indeed want *everything* inside `AbstactCommand` to be public. I would prefer just the `AbstractCommand` methods we expect to be invoked/implemented (`cmd_args` and `run`).

**Comment by dduugg** [comment]
Here's the deprecation notice when using https://github.com/DomT4/homebrew-autoupdate :

```
$ brew autoupdate start 43200
Warning: Calling `brew autoupdate'. This command needs to be refactored, as it is written in a style that is deprecated! Use inheritance from `Homebrew::AbstractCommand' ( see https://docs.brew.sh/External-Commands ) instead.
Please report this issue to the domt4/homebrew-autoupdate tap (not Homebrew/brew or Homebrew/homebrew-core), or even better, submit a PR to fix it:
  /opt/homebrew/Library/Taps/domt4/homebrew-autoupdate/cmd/autoupdate.rb:11

Homebrew will now automatically update every 12 hours.
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @dduugg!

File: Library/Homebrew/abstract_command.rb [hunk 2]
```diff
@@ -59,7 +64,15 @@ def initialize(argv = ARGV.freeze)
...
+    #
+    # @api public
     sig { abstract.void }
     def run; end
   end
+
+  module Cmd
+    # The command class for `brew` itself, allowing its args to be parsed.
+    class Brew < AbstractCommand; end
+  end
```
Comment (ZhongRuoyu) [inline Library/Homebrew/abstract_command.rb:77]: The addition of `Homebrew::Cmd::Brew` seems to have broken `brew typecheck --update` with:

```console
$ brew tc --update
Loading DSL extension classes... Done
Loading Rails application... Done
Loading DSL compiler classes... Done
Compiling DSL RBI files...

   identical  sorbet/rbi/dsl/homebrew/cmd/analytics.rbi
   identical  sorbet/rbi/dsl/homebrew/cmd/autoremove.rbi
Error: `Tapioca::Compilers::Args` failed to generate RBI for `Homebrew::Cmd::Brew`
   identical  sorbet/rbi/dsl/homebrew/cmd/cache.rbi
bundler: failed to load command: tapioca (/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/bin/tapioca)
/usr/local/Homebrew/Library/Homebrew/sorbet/tapioca/compilers/args.rb:34:in `decorate': undefined method `name' for nil (NoMethodError)
	from /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/tapioca-0.16.0/lib/tapioca/dsl/pipeline.rb:204:in `block in rbi_for_constant'
	from /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/tapioca-0.16.0/lib/tapioca/dsl/pipeline.rb:200:in `each'
	from /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/tapioca-0.16.0/lib/tapioca/dsl/pipeline.rb:200:in `rbi_for_constant'
	from /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/tapioca-0.16.0/lib/tapioca/dsl/pipeline.rb:88:in `block in run'
	from /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/gems/parallel-1.26.2/lib/parallel.rb:650:in `call_with_index'
[...]
```

Comment (dduugg) [reply Library/Homebrew/abstract_command.rb:77]: Thanks for flagging. I’m AFK but can have a fix in the morning (PDT) if no one beats me to it.

**Pull request merged**